aeron-driver/src/main/java/io/aeron/driver/DriverConductor.java

```diff
@@ -1968,21 +1972,54 @@ private <T extends DriverManagedResource> void checkManagedResources(
 
             resource.onTimeEvent(nowNs, nowMs, this);
 
+//            if (resource.hasReachedEndOfLife())
+//            {
+//                if (resource.free())
+//                {
+//                    fastUnorderedRemove(list, i, lastIndex--);
+//                    CloseHelper.close(ctx.errorHandler(), resource);
+//                }
+//                else
+//                {
+//                    ctx.systemCounters().get(FREE_FAILS).incrementOrdered();
+//                }
+//            }
+
             if (resource.hasReachedEndOfLife())
             {
-                if (resource.free())
-                {
-                    fastUnorderedRemove(list, i, lastIndex--);
-                    CloseHelper.close(ctx.errorHandler(), resource);
-                }
-                else
-                {
-                    ctx.systemCounters().get(FREE_FAILS).incrementOrdered();
-                }
+                CloseHelper.close(ctx.errorHandler(), resource::close);
+                endOfLiveResources.add(resource);
+                fastUnorderedRemove(list, i, lastIndex--);
             }
         }
     }
 
+    private int freeEndOfLifeResources(final int freeLimit)
+    {
+        int workCount = 0;
+
+        for (int i = 0; i < freeLimit; i++)
+        {
+            final DriverManagedResource resource = endOfLiveResources.pollFirst();
+            if (null == resource)
+            {
+                break;
+            }
+
+            if (resource.free())
+            {
+                workCount++;
+            }
+            else
+            {
+                ctx.systemCounters().get(FREE_FAILS).incrementOrdered();
+                endOfLiveResources.addLast(resource);
```

Comment:
This changes the order in which the resources are closed, is this intentional?

Reply:
This only changes the ordering of the `free` operations.  This is now independent of the `close`.  Frees should be able to happen in any order (they are just unmapping and removing files).  Closes are decoupled from the frees and have already happened in the appropriate order.